haystack/schema.py

```diff
@@ -549,23 +566,24 @@ def no_answer(self) -> Optional[bool]:
         return no_answer
 
     def to_dict(self):
-        return asdict(self)
+        return asdict(self, dict_factory=_dict_factory)
 
     @classmethod
-    def from_dict(cls, dict: dict):
-        # backward compatibility for old labels using answers with document_id instead of document_ids
+    def from_dict(cls, dict: "dict"):
         answer = dict.get("answer")
-        if answer and "document_id" in answer:
-            dict = dict.copy()
+        if answer and isinstance(answer, builtins.dict):
             dict["answer"] = Answer.from_dict(dict["answer"])
-        return _pydantic_dataclass_from_dict(dict=dict, pydantic_dataclass_type=cls)
+        doc = dict.get("document")
+        if doc and isinstance(doc, builtins.dict):
```

Comment:
Perfect change to use the [walrus operator](https://realpython.com/python-walrus-operator/)!
Also there's no need to check if `doc` is `None`, `isinstance` will do it for you.

```suggestion
        if isinstance(doc := dict.get("document"), Dict):
```

Note that I used `Dict` here instead of `dict`.